File: android/app/src/main/java/org/beetlebug/lookout/settings/AdvancedSection.kt
package org.beetlebug.lookout.settings

// The Advanced section: safety and quality, sizing, dates, and About.
//
// About sits at its foot, where the other shells put it: the version and the
// engine's pin are what a bug report needs, and the licences are a legal
// obligation that has to be reachable from somewhere a mariner can find.

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import org.beetlebug.lookout.licenses.LicenseManifest
import org.beetlebug.lookout.licenses.appVersion

@Composable
internal fun AdvancedSection(m: MarinerState, onOpenLicenses: () -> Unit) {
    SectionHeader("Safety & quality", first = true)
    SwitchRow("Data quality overlay", m, MI.DATA_QUALITY)
    SwitchRow("Isolated dangers in shallow water", m, MI.SHOW_ISOLATED_DANGERS_SHALLOW)
    SwitchRow("Information callouts", m, MI.SHOW_INFORM_CALLOUTS)
    SwitchRow("Meta boundaries", m, MI.SHOW_META_BOUNDS)
    SwitchRow("Overscale indication", m, MI.SHOW_OVERSCALE)

    SectionHeader("Sizing")
    SizeRow("Symbols & lines", m, MI.SIZE_SCALE)
    SizeRow("Text", m, MI.TEXT_SIZE_SCALE)
    SizeRow("Soundings", m, MI.SOUNDING_SIZE_SCALE)

    SectionHeader("Dates")
    SwitchRow("Date-dependent features", m, MI.DATE_DEPENDENT)
    SwitchRow("Highlight date-dependent", m, MI.HIGHLIGHT_DATE_DEPENDENT)
    Footer("Date-dependent features appear only when in season.")

    AboutSection(onOpenLicenses)
}

/**
 * What this build is, and the way to its terms. It sits at the foot of
 * Advanced, where the other shells put it: the version and the engine's pin
 * are what a bug report needs, and the licenses are a legal obligation that
 * has to be reachable from somewhere a mariner can find.
 */
@Composable
private fun AboutSection(onOpenLicenses: () -> Unit) {
    val manifest = LicenseManifest.current
    val engine = manifest?.components?.firstOrNull { it.id == "tile57" }
    val count = manifest?.components?.size ?: 0

    SectionHeader("About")
    ValueRow("Version", appVersion())
    if (engine != null && engine.pinLabel.isNotEmpty()) {
        ValueRow("Chart engine", "${engine.name} · ${engine.pinLabel}", mono = true)
    }
    Row(
        Modifier
            .fillMaxWidth()
            .clickable(onClick = onOpenLicenses)
            .padding(horizontal = 20.dp, vertical = 12.dp)
            .testTag("about-licenses"),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text("Licenses", style = MaterialTheme.typography.bodyMedium, modifier = Modifier.weight(1f))
        if (count > 0) {
            Text(
                "$count components",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        Icon(
            Icons.AutoMirrored.Filled.KeyboardArrowRight,
            contentDescription = null, // the row's own label says where it goes
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

File: apple/LookoutMarine/Settings/AdvancedSection.swift
//  AdvancedSection.swift — safety and quality, sizing, dates, and About.
//
//  Where anything unclaimed lands, and where the licences are reached from.

import SwiftUI


// MARK: - Advanced

struct AdvancedSections: View {
    @ObservedObject var m: MarinerSettings
    var body: some View {
        Section("Safety & Quality") {
            Toggle("Data quality overlay", isOn: $m.dataQuality)
            Toggle("Isolated dangers in shallow water", isOn: $m.showIsolatedDangersShallow)
            Toggle("Information callouts", isOn: $m.showInformCallouts)
            Toggle("Meta boundaries", isOn: $m.showMetaBounds)
            Toggle("Overscale indication", isOn: $m.showOverscale)
        }
        Section("Sizing") {
            SizeRow("Overall size", $m.sizeScale)
            SizeRow("Text size", $m.textSizeScale)
            SizeRow("Sounding size", $m.soundingSizeScale)
        }
        Section {
            Toggle("Date-dependent features", isOn: $m.dateDependent)
            Toggle("Highlight date-dependent", isOn: $m.highlightDateDependent)
            LabeledContent("View date") {
                TextField("YYYYMMDD", text: $m.dateView)
                    .frame(width: 110)
                    .multilineTextAlignment(.trailing)
            }
        } header: {
            Text("Dates")
        } footer: { Text("Leave the date empty to use today.").captionFooter() }
        AboutSection()
    }
}



private struct AboutSection: View {
    private var engine: LicenseComponent? {
        LicenseManifest.current?.components.first { $0.id == "tile57" }
    }

    private var componentCount: Int { LicenseManifest.current?.components.count ?? 0 }

    var body: some View {
        Section {
            LabeledContent("Version", value: LicensesView.appVersion)
                .monospacedDigit()
            if let engine, !engine.pinLabel.isEmpty {
                LabeledContent("Chart engine") {
                    Text("\(engine.name) · \(engine.pinLabel)")
                        .font(.system(.body, design: .monospaced))
                        .textSelection(.enabled)
                }
            }
            #if os(macOS)
            // The ellipsis is the platform's promise that a window opens.
            Button {
                LicensesWindowController.shared.show()
            } label: {
                LabeledContent("Licenses…", value: componentCount > 0 ? "\(componentCount) components" : "")
            }
            .buttonStyle(.plain)
            #else
            // The phone pushes the screen, so no ellipsis.
            NavigationLink {
                LicensesRoot()
            } label: {
                LabeledContent("Licenses",
                               value: componentCount > 0 ? "\(componentCount) components" : "")
            }
            #endif
        } header: {
            Text("About")
        }
    }
}



private struct SizeRow: View {
    let title: String
    @Binding var value: Double
    init(_ title: String, _ value: Binding<Double>) { self.title = title; self._value = value }
    var body: some View {
        VStack(alignment: .leading, spacing: 3) {
            HStack {
                Text(title)
                Spacer()
                Text(String(format: "%.2f×", value)).foregroundStyle(.secondary).monospacedDigit()
            }
            Slider(value: $value, in: 0.5...2.0)
        }
    }
}
